src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -361,6 +361,444 @@ BaseFloat DecodableAmNnetSimpleParallel::LogLikelihood(int32 frame,
   return decodable_nnet_->GetOutput(frame, pdf_id);
 }
 
+BatchNnetComputer::BatchNnetComputer(
+    const NnetSimpleComputationOptions &opts,
+    const Nnet &nnet,
+    const VectorBase<BaseFloat> &priors,
+    int32 online_ivector_period,
+    int32 minibatch_size):
+    opts_(opts),
+    nnet_(nnet),
+    output_dim_(nnet_.OutputDim("output")),
+    log_priors_(priors),
+    online_ivector_period_(online_ivector_period),
+    compiler_(nnet_, opts_.optimize_config),
+    minibatch_size_(minibatch_size) {
+  // Check the Nnet
+  KALDI_ASSERT(IsSimpleNnet(nnet));
+  ComputeSimpleNnetContext(nnet, &nnet_left_context_, &nnet_right_context_);
+  log_priors_.ApplyLog();
+  CheckAndFixConfigs();
+  //Prepare ComputationRequest and store a NnetComputation object
+  int32 input_dim = nnet_.InputDim("input"),
+        ivector_dim = nnet_.InputDim("ivector");
+  KALDI_ASSERT(input_dim > 0);
+  
+  int32 left_context = std::max(opts_.extra_left_context,
+                                opts_.extra_left_context_initial) +
+                       nnet_left_context_,
+        right_context = std::max(opts_.extra_right_context,
+                                 opts_.extra_right_context_final) +
+                       nnet_right_context_;
+  int32 num_input_rows = left_context + opts_.frames_per_chunk + right_context,
+        num_output_rows = opts_.frames_per_chunk / opts_.frame_subsampling_factor;
+  
+  
+  ComputationRequest request;
+  request.need_model_derivative = false;
+  request.store_component_stats = false;
+  request.inputs.reserve(2);
+  request.outputs.reserve(1);
+  std::vector<Index> input_indexes, ivector_indexes, output_indexes;
+
+  for (int32 n=0; n < minibatch_size_; n++) {
+    for (int32 t=0; t < num_input_rows; t++) {
+      input_indexes.push_back(Index(n, t - left_context, 0));
+    }
+  }
+  if (ivector_dim > 0) {
+    for (int32 n=0; n < minibatch_size_; n++) {
+      ivector_indexes.push_back(Index(n,0,0));
+    }
+  }
+  for (int32 n=0; n < minibatch_size_; n++) {
+    for (int32 i=0; i < num_output_rows; i++) {
+      output_indexes.push_back(Index(n, i * opts_.frame_subsampling_factor, 0));
+    }
+  }
+  
+  request.inputs.push_back(IoSpecification("input", input_indexes));
+  if (ivector_dim > 0) {
+    request.inputs.push_back(IoSpecification("ivector", ivector_indexes));
+  }
+  request.outputs.push_back(IoSpecification("output", output_indexes));
+  computation_ = compiler_.Compile(request);
+
+  num_batch_input_rows_ = num_input_rows;
+  num_batch_output_rows_ = num_output_rows;
+}
+
+
+void BatchNnetComputer::CheckAndFixConfigs() {
```

Comment:
This function is not specific to Batch decoding. Do we need to reimplement that?

Reply:
I explain it below. See it in the answer of "friend class" comment.